### ISSUE TITLE
# on the code of choose_action

### ISSUE THREAD

**Issue by HandsomeManKris** [JungangGe/DRL_for_DDBC#2]
Hi, Jungang

Thanks for sharing your code.

I have one question on your code as below:

 # epsilon greedy
        if random.uniform(0, 1) > self.epsilon:
            observation = observation[np.newaxis, :]
            actions_value = self.model1.predict(observation)
            action = np.argmax(actions_value)
        else:
            action = random.randint(0, self.n_actions - 1)
        return action
and code:

tar_nn = NeuralNetwork()

        eval_nn = NeuralNetwork()
        # 此时才是真有model
        self.model1 = tar_nn.get_model()
        self.model2 = eval_nn.get_model()
        # 两个 model 初始化相同的weights
        self.target_replace_op()
        # RMSProp optimizer
        optimizer = RMSprop(lr=self.lr, decay=self.lr_decay)
        self.model2.compile(loss='mse', optimizer=optimizer)
 
In my poor understanding, the model2 is train NN in paper and model1 is target NN in paper, but the problem is in your code, model1 (target NN) is used to choose action instead of model2(train NN). This makes me confused. Hope to get your reply on this question ~

**Comment by JungangGe** [comment]
Thanks for your question. I think that you are exactly right. The code for the DRL agent is actually originated from https://github.com/Jason33Wang/ReinforcementLearning_by_keras/blob/master/blogs.py, while I made some modifications. It now appears that there exsit a quite serious error in both Jason33Wang's code and my code. As the DQL part, I now suggest you to learn with the code at https://github.com/MorvanZhou/Reinforcement-learning-with-tensorflow/blob/master/contents/5_Deep_Q_Network/RL_brain.py. I am sorry to make you feel confused. 

Besides, I am now confused about why the code with this error can still works, both in Jason33Wang's code and in my simulation. I will try to check this.

**Comment by JungangGe** [comment]
> Hi, Jungang
> 
> Thanks for sharing your code.
> 
> I have one question on your code as below:
> 
> # epsilon greedy
> ```
>     if random.uniform(0, 1) > self.epsilon:
>         observation = observation[np.newaxis, :]
>         actions_value = self.model1.predict(observation)
>         action = np.argmax(actions_value)
>     else:
>         action = random.randint(0, self.n_actions - 1)
>     return action
> ```
> 
> and code:
> 
> tar_nn = NeuralNetwork()
> 
> ```
>     eval_nn = NeuralNetwork()
>     # 此时才是真有model
>     self.model1 = tar_nn.get_model()
>     self.model2 = eval_nn.get_model()
>     # 两个 model 初始化相同的weights
>     self.target_replace_op()
>     # RMSProp optimizer
>     optimizer = RMSprop(lr=self.lr, decay=self.lr_decay)
>     self.model2.compile(loss='mse', optimizer=optimizer)
> ```
> 
> In my poor understanding, the model2 is train NN in paper and model1 is target NN in paper, but the problem is in your code, model1 (target NN) is used to choose action instead of model2(train NN). This makes me confused. Hope to get your reply on this question ~

Thanks for your question. I think that you are exactly right. The code for the DRL agent is actually originated from https://github.com/Jason33Wang/ReinforcementLearning_by_keras/blob/master/blogs.py, while I made some modifications. It now appears that there exsit a quite serious error in both Jason33Wang's code and my code. As the DQL part, I now suggest you to learn with the code at https://github.com/MorvanZhou/Reinforcement-learning-with-tensorflow/blob/master/contents/5_Deep_Q_Network/RL_brain.py. I am sorry to make you feel confused.

Besides, I am now confused about why the code with this error can still works, both in Jason33Wang's code and in my simulation. I will try to check this.

**Comment by HandsomeManKris** [comment]
Thank you for your reply
Don't worry man, your code is really good and give a me a lot of help on deep reinforcement learning.
BTW
There is another place at the steering vector below:
    def _generate_steering_vector_(self):
        # random.rand: get variable in [0,1) with 1X4 vector 这个地方似乎有错误
        # self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 0.25) * 2 * self.angular_spread
        self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 1) * self.angular_spread
        self.sv = np.zeros((self.multi_paths, self.bs.n_antennas), dtype=complex)
        for i in range(self.multi_paths):
            self.sv[i, :] = np.exp(1j * np.pi * np.cos(self.aod[i]) * np.arange(self.bs.n_antennas)) \
                              / np.sqrt(self.bs.n_antennas)
the factor of agular spread should be uniformly distributed between [-1/2, 1/2] and your code seems to generate a vector with each element uniformly distributed between [-1, -1/2], I made a modification below:

self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 0.25) * 2 * self.angular_spread

:)
In my poor opinion, all other codes are correct and I'd like to have more touch with you in the future. Your paper is so good!!!

**Comment by JungangGe** [comment]
> Thank you for your reply
> Don't worry man, your code is really good and give a me a lot of help on deep reinforcement learning.
> BTW
> There is another place at the steering vector below:
> def _generate_steering_vector_(self):
> # random.rand: get variable in [0,1) with 1X4 vector 这个地方似乎有错误
> # self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 0.25) * 2 * self.angular_spread
> self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 1) * self.angular_spread
> self.sv = np.zeros((self.multi_paths, self.bs.n_antennas), dtype=complex)
> for i in range(self.multi_paths):
> self.sv[i, :] = np.exp(1j * np.pi * np.cos(self.aod[i]) * np.arange(self.bs.n_antennas))
> / np.sqrt(self.bs.n_antennas)
> th�e factor of agular spread should be uniformly distributed between [-1/2, 1/2] and your code seems to generate a vector with each element uniformly distributed between [-1, -1/2], I made a modification below:
> 
> self.aod = self.norminal_aod + (0.5 * np.random.rand(self.multi_paths) - 0.25) * 2 * self.angular_spread
> 
> :)
> In my poor opinion, all other codes are correct and I'd like to have more touch with you in the future. Your paper is so good!!!

Thanks a lot for pointing out the errors in my code.  I am very glad to see that my code could help you.

As for the error you pointed in second comment, just use "np.random.rand(self.multi_paths)-0.5" is OK,  I think. 

Thanks again for your carefulness in reading my code. Wish you success in your research.
